src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Helpers.cs

```diff
@@ -208,6 +208,35 @@ public static void ValidateDer(ReadOnlyMemory<byte> encodedValue)
                 reader.ReadEncodedValue();
             }
         }
+
+        public static ReadOnlyMemory<byte> DecodeOctetStringAsMemory(ReadOnlyMemory<byte> encodedOctetString)
+        {
+            AsnReader reader = new AsnReader(encodedOctetString, AsnEncodingRules.BER);
+
+            if (reader.PeekEncodedValue().Length != encodedOctetString.Length)
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            // Almost everything in X.509 is DER-encoded, which means Octet String values are
+            // encoded as a primitive (non-segmented)
+            //
+            // Even in BER Octet Strings are usually encoded as a primitive.
+            if (reader.TryReadPrimitiveOctetStringBytes(out ReadOnlyMemory<byte> primitiveContents))
+            {
+                return primitiveContents;
+            }
+
+            byte[] tooBig = new byte[encodedOctetString.Length];
```

Comment:
is this mostly 1:1 actual to estimated size or can there be very large difference? (except for maliciously crafted payloads)

Reply:
In CER, it's 3 bytes overhead per 1000 bytes of content.  So if it wasn't the early return the expected state is 99.7% filled.